File: cocos-creator-project/assets/scripts/game/IFFSystem.ts
import { AircraftType, IFFStatus, GameConstants } from './GameTypes';
import { Aircraft } from './Aircraft';

// ============================================================================
// IFF (Identification Friend or Foe) System
// Automatic interrogation of radar contacts using Mode 4
// ============================================================================

interface IFFInterrogation {
    trackId: number;
    timeRemaining: number;
    aircraft: Aircraft;
}

export class IFFSystem {
    private pendingInterrogations: IFFInterrogation[] = [];
    private errorRate: number = 0;

    interrogate(aircraft: Aircraft): void {
        if (!aircraft || aircraft.getTrackId() < 0) return;
        if (this.isInterrogating(aircraft.getTrackId())) return;
        if (aircraft.getIFFStatus() !== IFFStatus.PENDING) return;

        this.pendingInterrogations.push({
            trackId: aircraft.getTrackId(),
            timeRemaining: GameConstants.IFF_INTERROGATION_TIME,
            aircraft,
        });
    }

    update(dt: number): void {
        this.pendingInterrogations = this.pendingInterrogations.filter((interrog) => {
            interrog.timeRemaining -= dt;

            if (interrog.timeRemaining <= 0) {
                if (interrog.aircraft && interrog.aircraft.isAlive()) {
                    const result = this.determineResult(interrog.aircraft);
                    interrog.aircraft.setIFFStatus(result);
                }
                return false; // remove completed
            }
            return true; // keep pending
        });
    }

    isInterrogating(trackId: number): boolean {
        return this.pendingInterrogations.some((i) => i.trackId === trackId);
    }

    setErrorRate(rate: number): void {
        this.errorRate = rate;
    }

    private determineResult(aircraft: Aircraft): IFFStatus {
        const actuallyFriendly = aircraft.isFriendly();

        if (Math.random() < this.errorRate) {
            const errorType = Math.random();
            if (errorType < 0.3) {
                return actuallyFriendly ? IFFStatus.HOSTILE : IFFStatus.FRIENDLY;
            }
            return IFFStatus.UNKNOWN;
        }

        if (aircraft.getType() === AircraftType.STEALTH_FIGHTER) {
            if (Math.random() < 0.3) {
                return IFFStatus.UNKNOWN;
            }
        }

        return actuallyFriendly ? IFFStatus.FRIENDLY : IFFStatus.HOSTILE;
    }
}

File: src/IFFSystem.cpp
#include "IFFSystem.h"
#include <random>
#include <algorithm>

IFFSystem::IFFSystem()
    : errorRate_(0.0f)
{
}

IFFSystem::~IFFSystem() {}

void IFFSystem::interrogate(Aircraft* aircraft)
{
    if (!aircraft || aircraft->getTrackId() < 0) return;

    // Don't interrogate if already being interrogated
    if (isInterrogating(aircraft->getTrackId())) return;

    // Don't re-interrogate if already has a result
    if (aircraft->getIFFStatus() != IFFStatus::PENDING) return;

    IFFInterrogation interrog;
    interrog.trackId = aircraft->getTrackId();
    interrog.timeRemaining = GameConstants::IFF_INTERROGATION_TIME;
    interrog.aircraft = aircraft;

    pendingInterrogations_.push_back(interrog);
}

void IFFSystem::update(float dt)
{
    auto it = pendingInterrogations_.begin();
    while (it != pendingInterrogations_.end()) {
        it->timeRemaining -= dt;

        if (it->timeRemaining <= 0.0f) {
            // Interrogation complete - determine result
            if (it->aircraft && it->aircraft->isAlive()) {
                IFFStatus result = determineResult(it->aircraft);
                it->aircraft->setIFFStatus(result);
            }
            it = pendingInterrogations_.erase(it);
        } else {
            ++it;
        }
    }
}

bool IFFSystem::isInterrogating(int trackId) const
{
    for (const auto& interrog : pendingInterrogations_) {
        if (interrog.trackId == trackId) return true;
    }
    return false;
}

IFFStatus IFFSystem::determineResult(Aircraft* aircraft)
{
    static std::mt19937 rng(std::random_device{}());
    std::uniform_real_distribution<float> dist(0.0f, 1.0f);

    bool actuallyFriendly = aircraft->isFriendly();

    // Check for IFF error
    if (dist(rng) < errorRate_) {
        // Error! Return incorrect or inconclusive result
        float errorType = dist(rng);
        if (errorType < 0.3f) {
            // Complete misidentification (dangerous!)
            return actuallyFriendly ? IFFStatus::HOSTILE : IFFStatus::FRIENDLY;
        } else {
            // Inconclusive
            return IFFStatus::UNKNOWN;
        }
    }

    // Stealth aircraft have a chance of appearing as UNKNOWN
    if (aircraft->getType() == AircraftType::STEALTH_FIGHTER) {
        if (dist(rng) < 0.3f) {
            return IFFStatus::UNKNOWN;
        }
    }

    // Normal operation - correct identification
    return actuallyFriendly ? IFFStatus::FRIENDLY : IFFStatus::HOSTILE;
}
